triq/triq.py

```diff
@@ -1,30 +1,71 @@
+'''
+Ideas how to use trio guest mode with Qt borrowed from:
+
+https://github.com/altendky/alqtendpy/pull/19/files
+'''
 import trio
-from collections import deque
+import traceback
+try:
+    from PyQt5.QtCore import QEvent, QObject
+except ImportError:
+    from PySide2.QtCore import QEvent, QObject
 
-_tasks = deque()
 
-_exit_event = trio.Event()
+class ReenterEvent(QEvent):
+    REENTER_EVENT = QEvent.Type(QEvent.registerEventType())
 
-_TIMING = 0.005
+    def __init__(self, fn):
+        self._fn = fn
+        super().__init__(self.REENTER_EVENT)
+
+    def consume(self):
+        self._fn()
+        return False
+
+class Reenter(QObject):
+    def event(self, event):
+        return event.consume()
+
+
+_MAX_CONCURRENT_TASKS = 100
+_send_channel, _receive_channel = trio.open_memory_channel(_MAX_CONCURRENT_TASKS)
+
+_EXIT = object()
 
-def call_async(fn, *av):
-    _tasks.append( (fn, av) )
 
 def exit():
-    _exit_event.set()
+    _send_channel.send_nowait(_EXIT)
 
-async def _main(app):
-    async with trio.open_nursery() as nursery:
-        while not _exit_event.is_set():
-            app.processEvents()
-            if _tasks:
-                fn, args = _tasks.popleft()
-                nursery.start_soon(fn, *args)
-            else:
-                with trio.move_on_after(_TIMING):
-                    await _exit_event.wait()
+
+def call_async(fn, *av):
+    _send_channel.send_nowait( (fn, av) )
 
 
 def run(app):
-    trio.run(_main, app)
-    app.exit()
+    _reenter = Reenter()
+
+    async def _main():
+        with trio.CancelScope() as cancel_scope:
+            app.lastWindowClosed.connect(cancel_scope.cancel)
+            async for x in _receive_channel:
+                if x is _EXIT:
+                    break
+                try:
+                    fn, av = x
+                    await fn(*av)
```

Comment:
I had been focusing on a Trio-centric system and was building up ways for Trio to wait on signals and such.  I hadn't gotten to the Qt->Trio side of things yet.  This memory channel approach seems reasonable at least with the minute of thought I've given it so far.  :]

Instead of `await`ing here, should this be a `nursery.start_soon()` to avoid blocking other tasks from being initiated as well?